### DISCUSSION TITLE
Windows 98 running on Dosbox Staging (Tips & Tricks + Findings)

### DISCUSSION THREAD

**Discussion by ThomasEricB** [dosbox-staging/dosbox-staging#2054] [General]
Hello folks,

As I was made aware by @kcgen on Discord, Windows 98 and 95 is outside of the scope of the project... Nonetheless, I got it working with the [Fat32 test](https://github.com/dosbox-staging/dosbox-staging/tree/ww/fat32-test) (see #1964 ). Here some observations I had:

- DOSBox Staging gives a lot of errors if Win98 has any startup programs. So what I did was going on DOSBox-x and disabling it inside those programs.
- Windows 98 does not recognize any display settings inside DOSBox Staging, so what I did was altering them on DOSBox-x.
- I would recommend setting `mcb_fault_strategy = repair` because in my testing it made it run (reduced the crashing close to 0)
- I would leave `machine = svga_s3` as loading anything above that (svga_et3000, svga_et4000, svga_paradise, vesa_nolfb, vesa_oldvbe) would instantly crash Windows 98
- `vesa_modes = compatible` and `vmemsize = 8` works great!
- `startup_verbosity  = auto` is recommended
- `cputype = pentium_slow` is recommended (though 486_slow also works and others probably will)
- `core = dynamic` is recommended
- EDIT 4: **I do not recommend setting a fixed number of cycles for this.** In my testing setting any amount caused a bunch of issues for me. However, keep in mind it might be my machine or my setup causing this issue. Either way, I strongly recommend setting your .conf file with either `cycles = max 105%` or `cycles = max` or `cycles = auto`.
- EDIT 2: I don't recommend moving the mouse during Startup (+ crashes) - that has increased the amount of errors for me. 
- EDIT 3: Might be a placebo, but I recommend letting ScanDisk run if you had the notification in EDIT 1 on any app (below)
- EDIT 1 (as I forgot to add): Close [Kill] and Reopen Dosbox Staging if you get any notification like this (DO NOT CLICK OR INTERACT WITH IT): 

![dosbox_zSkfJYDrs8](https://user-images.githubusercontent.com/9662095/198874174-70e2b374-eb74-416f-90c9-26e9b4a712b3.png)

Besides that, here is the autoexec I was running:

```
@echo off
mount c: e:\test\
imgmount 0 E:\test\win98c.img -t floppy -fs none
imgmount 2 E:\test\Win98SE.img -size 512,63,255,522 -t hdd -fs none
imgmount e E:\dosbox-staging-windows-x86_64-v0.79.1\dosbox-staging-windows-x86_64-v0.79.1\diablo.iso -t iso -fs none
COUNTRY 55 #sets country to Brazil
KEYB br275 #sets keyboard to ABNT Brazil
boot c:
```

[And here are the results on DOOM!](https://www.youtube.com/watch?v=t5d1qHSeTBA):

[![Watch the video](https://img.youtube.com/vi/t5d1qHSeTBA/maxresdefault.jpg)](https://youtu.be/t5d1qHSeTBA)

**Comment by ThomasEricB** [comment]
Yikes, I just found a nasty bug. Moving your mouse too fast causes Windows 98 to freeze and if you continuously do so will make it crash. See: https://youtu.be/IRprhj5nW-A?t=172 for the exact time of the crash. See entire video to confirm it was running.

**Reply by ThomasEricB** [reply]
Created issues #2056 and https://github.com/joncampbell123/dosbox-x/issues/3801 about this.

**Reply by Torinde** [reply]
- #4501

**Comment by arrowgent** [comment]
fairly similar situation for win95

while it kind of works, there are limitations that make using win9x not reasonable in dosbox

**Comment by symbios24** [comment]
Hello

Im trying to run games like age of empires 2
In windows 95 and 98 using a raspberry pi 4 and Dosbox-x but they are running very slow,i get 5fps in Doom95, I've seen videos in youtube running age of empires 2 and other games very smooth in Raspberry Pi 4 in Dosbox svn and Dosbox ece but cannot install DOSBox-svn and ece successfully they crasyhing in first boot with explorer.exe and dll errors,any help here???

Thank you

**Reply by ThomasEricB** [reply]
???? I have no videos of the sort nor this is Dosbox-X

**Reply by symbios24** [reply]
Not your videos

https://m.youtube.com/watch?v=lYmfklwWK2k

https://m.youtube.com/watch?v=W5vt6jtzhlg

**Reply by symbios24** [reply]
I said maybe you know how to make dosbox run in this speed that games

**Comment by almeath** [comment]
Most of this guide appears to hold true in the latest build of Staging (0.81.0). I had already worked out most of the above settings based on extensive testing with DOSBox SVN. Personally, I find "mcb_fault_strategy = allow" gives me better results than repair. I was getting a lot of crashes in File Explorer until I made that change.

I notice the autoexec bat contains a line to mount an ISO file as E: drive, but that has no effect in the Windows environment because there is currently no IDE emulation in DOSBox Staging, which is one of the reasons DOSBox-X has an advantage in this area. However, the raw speed of the dynamic core in the latest 0.81.0 alpha build does make up for this with demanding games, particularly those requiring 3DFX. 

When I was using a Mac, I patched the DOSBox Pure implementation of the 3DFX emulation into DOSBox SVN, and it worked great. Much more stable and speedy, even with bilinear filtering enabled. I found the implementation of the older 3DFX patch in DOSBox-X to be very slow. I am happy that the Staging crew have seemingly taken the path of implementing the DOSBox Pure version of the patch into the current code base.

If only there was IDE emulation. That is the last remaining hurdle to making Staging a better option than DOSBox-X, at least for the purpose of playing games in a Win9x environment. I am sure DOSBox-X would still be superior for offering stability and a host of refinements to more properly support a fully functioning Win9x install.

So we now have:

1. sufficient dynamic core performance = check
2. Support for large RAM allocations = check
3. Support for large hard drive images = check
 
Missing IDE emulation means that using software like Virtual Clone drive is necessary for installing and running games. It also means that rare Windows-only mixed mode / Redbook audio CD games (i.e. Silent Thunder) still need DOSBox-X if you want to hear the music. I have tried every image-mounting program that ever worked in Win98 and none of them seem capable of reading Rebook audio off a bin/cue image.

**Reply by shermp** [reply]
@almeath have you figured out how to get the 12MB variant of Voodoo working in Win 98?

**Reply by shermp** [reply]
Also, IDE works (to an extent) for CD-ROM images. Is this not sufficient?

Actually, I've just tested with Descent 2 CD. It is not stable unfortunately :(

**Reply by almeath** [reply]
I need help understanding what the 3dfx patch is actually emulating. Earlier versions of the patch (i.e. the one used in DOSBox-X) state that it emulates a "3dfx Voodoo", i.e. not a Voodoo 2. When I look up specifications on the original Voodoo card, it seems that 4MB was standard. I am not aware if a 12MB variant was ever released? All references to 12MB are made in relation to the later Voodoo 2 card. I see that DOSBox Staging supports 12MB of ram, but does that turn it into a Voodoo 2, or only provide more RAM than was allowed on the original hardware? Lastly, when Voodoo 2 is enabled in 86Box or PCem, loading the correct drivers will provide the extra settings panel in the Windows Display control panel, which you do not get with Voodoo 1 drivers (as far as I know).

On the subject of IDE and CD-ROM mounting within Windows 98, I tried the command suggested in the first post above, i.e.:

imgmount 2 harddisk\hdd.img -size 512,63,32,1015 -t hdd -fs none
imgmount d media\Star Wars Rebellion.iso -t iso -t iso -fs none
boot -l c

I only see the empty "D" drive as CD-ROM.

I also tried changing the "d" to "3" and got the same result.

What command are you using to mount an ISO within Windows directly? 

I have been using Daemon Tools and Virtual Clone Drive to use CD images, which necessitates placing the ISO files on large hard drive images, which is a clumsy but usable work-around.

**Reply by shermp** [reply]
Dosbox Staging implemented (some of) the IDE emulation from X, primarily for CD ROM support. You can mound an image like so:

`imgmount d "C:\Games\dosbox\Descent2\cd\DESCENT_II.cue" -ide 2m -t cdrom`

Note the `-ide 2m`. This is the instruction to mount as the `master` drive on the secondary IDE controller. You can also mount a second drive to `-ide 2s`.

The drive should then show in Windows.

Unfortunately, I've noticed a few bugs. First, playing CD audio seems to freeze or outright crash Dosbox. Second, disk swapping doesn't seem to work properly.

Also note, to mount mixed mode CD's with data and audio tracks (like many 90s games), you must use a format like BIN/CUE. ISO does NOT work with such CD's.

As far as the Voodoo goes, I've been doing a bit of research, it *seems* there were a few so-called "Obsidion" cards that tacked on a second TMU to turn them into a Voodoo2 like card (but they were not Voodoo 2's)

**Reply by Torinde** [reply]
> When you select the 12 MB setting, it creates this "super Voodoo 1" board with dual TMUs.

https://github.com/dosbox-staging/dosbox-staging/pull/3103#issuecomment-1802814068

Hopefully [Voodoo2 support](https://github.com/dosbox-staging/dosbox-staging/issues/3115#issuecomment-1809675044) will be added, thus reducing the confusion.

**Reply by almeath** [reply]
I wonder what I am doing wrong.. 

I tried this:

imgmount 2 harddisk\hdd.img -size 512,63,64,1023 -t hdd -fs none
imgmount d media\Windows98.iso -ide 2m -t cdrom
boot -l c

And I see no secondary drive, CD-ROM or otherwise, when booting into Windows 98. I also tried the "-ide 2s" variant, and same thing. 

I even tried a fresh install of Windows 98 with the same result.

This is strange. Are there any configuration file settings that could prevent this from working? The ISO I am using is fine, it mounts outside of DOSBox Staging.

**Reply by shermp** [reply]
Only thing I can think of is that your Windows installation doesn't have any IDE controller drivers installed? I set up my Win 98 from the CD connected via IDE, and it shows up fine for me.

**Reply by almeath** [reply]
That might be it. I followed the traditional SVN approach of copying the "Win98" folder onto a blank disk image and installing from that. I will see if I can re-install using a mounted ISO.

**Reply by johnnovak** [reply]
What @Torinde and @shermp said. We only emulate the Voodoo 1, and these are just various configuration options. Different companies did different 3rd party Voodoo board implementations.

Voodoo 2 is pointless for DOS emulation, as nothing supports it. So I don't see why we should even bother.

**Reply by almeath** [reply]
Coming from SVN, I am new to this approach of installing Windows 98 directly from an ISO, and need a bit more help.

I have this in my conf file:

imgmount a media\WIN98C.IMG -t floppy
imgmount 2 harddisk\hdd.img -size 512,63,64,1023 -t hdd -fs none
imgmount d media\Windows98SE.iso -ide 2m -t cdrom

What would be the correct boot command to allow Windows to install directly off the ISO onto the (already formatted) C drive image?

When I launch Staging with only the above commands, I can see the contents of A and D, but C is not detected.

If I boot off the floppy image with "boot media\WIN98.IMG" then I can see the ISO file as R drive, but C drive is still not detectable.

**Reply by shermp** [reply]
```ini
[autoexec]
imgmount 2 "C:\DiskImages\db-staging-98se.img" -fs none -t hdd -size 512,63,255,1023
imgmount d "C:\DiskImages\MS\Win98.iso" -ide 2m -t cdrom
# boot -l c
boot "C:\DiskImages\MS\Win98SE-BootDisk.img"
```

This is what I start with. First boot with a 98 boot disk. It should give you the option to start setup from CD, or boot with CD enabled. I usually choose the second option, and run `fdisk` to create my disk partition. Then I restart dosbox, and choose the option to start setup from the CD.

From there, the Win 98 setup *should* find the hard disk, and prompt to format it.

Windows 98 will then proceed to do the initial installation, and copy files across. Once it wants to restart the first time, close Dosbox, and comment the boot disk line, and uncomment the `boot -l c` line, and start Dosbox again. Setup should continue.

Note that subsequent reboots will hang at the "Windows is shutting down" screen. Just exit dosbox, and start it again to continue.

**Reply by almeath** [reply]
Thanks, that did allow me to install off the CD iso.

When I have a CD iso image mounted, Windows gives me a black screen for about 10 seconds between the Windows loading screen and the desktop appearing, but then everything proceeds to run normally. If I only have a hard disk image mounted, and no CD iso, then there is no black screen or delay. Weird issue, but not a show stopper. When I load images from Daemon tools and similar applications, the black screen issue does not occur , so it appears to be caused by something specific to having a CD image actively mounted via the IDE emulation when Windows 98 starts.

Also, I tried mounting a bin/cue game with Redbook audio (Nightmare Creatures) and it freezes when the first level is about to start, but I can hear the music starting to play - essentially it totally freezes and I have to force quit DOSBox.

Interestingly, when I load the same bin/cue pair on my earlier Windows 98 install (the one I brought across from SVN without the IDE CD-ROM support) with Daemon tools 3.47 and a hacked version of the SB16 drivers (allowing digital audio to be enabled), the game runs fine with both 3dfx and the Redbook audio working.

So, I guess what I am saying is, that the ability to mount ISO and bin/cue directly is promising, but it is definitely showing signs of being unstable. However I do need to conduct further tests.

**Comment by FeralChild64** [comment]
More is missing regarding Windows 9x support, comparing to DOSBox-X. The CPU core does not handle page faults correctly enough which might lead to emulator crash, S3 card emulation is behind X and you have to be careful which driver you install, lack of FAT32 support makes data exchange less convenient, etc.

**Reply by rderooy** [reply]
There are some other areas that Staging is behind on for running Win9x. PCI, PnP, APM are some that come to mind.

**Reply by Torinde** [reply]
Related: https://github.com/dosbox-staging/dosbox-staging/pull/3426#issuecomment-1962846460, https://github.com/dosbox-staging/dosbox-staging/pull/3687#issuecomment-2132052993, https://github.com/dosbox-staging/dosbox-staging/discussions/1110#discussioncomment-6710874, #3115, https://github.com/dosbox-staging/dosbox-staging/issues/3955#issuecomment-2372633299, #4187, #4231, #3602, https://github.com/dosbox-staging/dosbox-staging/pull/4513#issuecomment-3302939705 ([Guest additions](https://github.com/JHRobotics/softgpu/issues/14) functionality), [6 axis 10 button joysticks](https://github.com/dosbox-staging/dosbox-staging/issues/3075#issuecomment-3435939619)

**Comment by almeath** [comment]
Yeah, not disputing any of that. DOSBox-X has for years focused on the odds and ends needed to better support Win9x. My objective has been firmly set on maximizing raw performance for games within the Win9x environment, and I found that for the last three years or so, up to date SVN was stable enough for most to work fine - I believe this has carried through into Staging. My day-to-day experience has been very little crashing once a game is running. Any instability, errors, sudden quits etc. have all happened when trying to do various things within the File Explorer. If key features are missing, most games do not seem to need them in order to run.

**Reply by johnnovak** [reply]
@almeath That's super informative and makes me want to play with Win98 in Staging a bit more 😎 At the end of the day, 99% of the time it's about the games.

**Reply by almeath** [reply]
Thanks @johnnovak , I feel I have been experimenting with this long enough that I am comfortable to share some insights, realizing that Win9x is not the focus of DOSBox. 

One interesting thing I have noticed, is that the Windows 98 SE installs I have used within DOSBox-X and SVN, when brought across to 0.81.0 (not not 0.80.x) will, upon booting, find the device "PCI Early non-VGA device" and then fail to find any suitable drivers on standard Windows 98 SE cd images. I thought it could be something to do with Voodoo drivers, but it also fails to find suitable drivers from the (already installed) set in Voodoo driver kit version 3.01.00. 

Anyway, it does not seem to do any harm, or stop anything working, whether reliant on 3dfx or not, as the Voodoo device stills shows up separately in the Device Manager. Whatever the cause, it is something triggered only in 0.81.0.

![Screenshot 2023-12-23 184358](https://github.com/dosbox-staging/dosbox-staging/assets/5726214/597fb756-ef8d-42fc-85f8-a94b023c427d)

Something I will get around to, is trying a completely fresh Windows 98 SE install within the latest Staging, and see if the issue persists.

**Reply by rderooy** [reply]
Do you perhaps have reelmagic enabled? In any case, assuming you can reproduce it, you may want to `git bisect` it.

**Reply by almeath** [reply]
Thanks, I will try the fresh installation first, to see if that clears it. One thing I can confirm is that it only affects Windows 98. Windows 95 installs with the same DOSBox config settings are unaffected. Also, Reelmagic is disabled.

**Reply by shermp** [reply]
I just discovered this yesterday on a fresh Windows 98 build (the extra device). 

On a side note, I tried `mcb_fault_strategy = repair` for the first time yesterday, and it was noticeably more unstable then when I had previously used `allow`.

**Reply by almeath** [reply]
I entirely agree - crashes all over the place when using repair, but allow is about 90% stable.

**Reply by almeath** [reply]
I tried a fresh install of Windows 98, and it comes up the same ; "PC Early Non-VGA Device", and it keeps re-detecting on boot even if I disable Voodoo and ReelMagic on the configuration settings. As reported earlier, it does not happen in SVN and DOSBox-X, so it is peculiar to DOSBox Staging 0.81.0

**Reply by johnnovak** [reply]
> I entirely agree - crashes all over the place when using repair, but allow is about 90% stable.

That's some useful info.

**Comment by mattisengstedt** [comment]
Does this work with the latest main release? I'm having issues.

**Reply by almeath** [reply]
Not sure which part of this topic you are referring to - everything described above still holds true under the latest release, as far as I know. All my Windows 98 configurations under 0.81.1 are performing the same as under previous releases. What is the issue you are experiencing?

**Reply by mattisengstedt** [reply]
I created a hdd in dosbox-x, mounted it in staging along with 98 se install iso and 98 se boot floppy. Deleted the partition in fdisk and started up the install. Formatted the hdd in the setup. After that scandisk found errors and had to abort.

**Reply by mattisengstedt** [reply]
Tried creating the partition and prepping it in fdisk instead, that works and I´ve started the install but it turned my 60gb hdd image into an 8gb one.

**Reply by almeath** [reply]
I know that a large hard drive / FAT 32 patch was implemented in DOSBox Staging, but I am not sure what the usable drive size limit is at the moment. I have never got more than 8GB working with Win 9x installations.

Can you test with a smaller drive size (i.e. 4GB) and see if it reacts the same? Also, I still have to specify drive geometry in my conf files, as far as I know, automatic geometry detection only works in DOSBox-X.

**Comment by whbinder** [comment]
Sorry to dig up this thread, but I can't get ANY HDD image to boot into Staging. I have images that can boot in 86Box, DOSBox-X, and PCem, but I get 

partition signature != 55aa

imgmount a Windows95b.img -t floppy
imgmount 2 "win3.img" -fs none -size 512,63,64,520
imgmount d cd.iso  -t cdrom -fs none

boot -l C:

If I boot into A: I can read the drive jsut fine, but booting can't find the MBR

Interestingly the actual Staging instructions includes this as an example
imgmount c win3.img -t hdd -fs none -size 512,63,64,520

But when actually putting that in my .conf it says "Must specify a Drive Number"

Any thoughts welcome.

**Comment by FeralChild64** [comment]
> But when actually putting that in my .conf it says "Must specify a Drive Number"

There is a mistake in the help - it should be a drive number, 0, 1, 2, or 3 if I'm not mistaken. If I remember correctly, I have created my images using 86Box - just make sure these are raw images, not some kind of VHD file.

Currently @dividebysandwich is working on improving the disk image support, hopefully you'll see improvements in 0.83.0.

**Reply by dividebysandwich** [reply]
Precisely, if you want to boot, the argument should be a drive number (A=0, B=1, C=2, D=3).
Though good point - why is that the case? Maybe we could make this work with drive names across the board and do the number resolution internally...

**Reply by johnnovak** [reply]
> Precisely, if you want to boot, the argument should be a drive number (A=0, B=1, C=2, D=3). Though good point - why is that the case? Maybe we could make this work with drive names across the board and do the number resolution internally...

Yeah, it seems silly. We shluld be able to use drive letters.

**Reply by FeralChild64** [reply]
> Yeah, it seems silly. We shluld be able to use drive letters.

Right, but if you want to boot a real OS, you need to connect the image to one of the 4 channels (primary/secondary IDE, master/slave channel). And I'm not sure if Windows 9x is totally happy if the drives change places from boot to boot - so, IMHO, the user should have an option to control to which controller/channel attach the image.

**Reply by johnnovak** [reply]
Yeah cool, but we don't even support Win9x... yet. I guess it will be handy for future Win9x support, if that ever happens.

**Reply by rderooy** [reply]
> > Yeah, it seems silly. We shluld be able to use drive letters.
> 
> Right, but if you want to boot a real OS, you need to connect the image to one of the 4 channels (primary/secondary IDE, master/slave channel). And I'm not sure if Windows 9x is totally happy if the drives change places from boot to boot - so, IMHO, the user should have an option to control to which controller/channel attach the image.

For this purpose, the `imgmount` command on DOSBox-X supports an `-ide` option which can be set to 1m,1s,2m,2s. For Win95 in particular you do not want the IDE HDD and IDE CD-ROM on the same controller, so you would mount the HDD with `-ide 1m` and the CD-ROM with `-ide 2m`.
